### DISCUSSION TITLE
Plugin Idea: SEO Auditing Tool

### DISCUSSION THREAD

**Discussion by nickreese** [Elderjs/elderjs#177] [Plugin Ideas]
# Update: Plugin is live

Plugin is live here: https://github.com/Elderjs/plugins/tree/master/packages/seo-check

--------
Hey all, I've got a script I've built that does SEO auditing for common issues. 

It checks HTML for things like:

* Missing title tag
* Missing meta description
* Missing canonical
* Canonical that doesn't match permalink
* Missing h tags in hierarchy. (h1 h3 h4 h5)
* And the list goes on. 

# Requests:

## Upvote If:

1. Please vote up this plugin idea if you'd like this to be an official plugin. 

## Comment With: 
1. Please comment below if you'd be willing to test this plugin.
2. Please comment below if you have SEO checks you'd like to see.
3. Please comment below if you'd like to help contribute to development.

**Comment by nickreese** [comment]
Other items my current implementation checks:
- Several other htag rules.
- Internal links are well formed. (trailing slash `/`)
- Internal links are lower case.
- Page has `<meta name="viewport" content="width=device-width, initial-scale=1.0" />`
- Meta description length isn't too long or to short.
- Meta description includes 20% of the words of the title tag.  (arbitrary but useful in my experience.)


--

Items I'd like to add:
- check for internal no follows
- Check for open graph tags and images.
- Check for HTML lang
- Check that internal links are all SSG'd pages or links to valid routes. 
- All images have alt tags. 

Long shot items:
- Ensure all pages have unique title tags
- Ensure all pages have unique descriptions
- Check for orphaned pages.

**Comment by versysmedia** [comment]
Definitely would be a cool addition

**Comment by kinooyume** [comment]
Seems a must have as elder position itself as an SEO focus framework. 

However, it seems that most of the features you mention concern the html file and can be use in any web project, so it feels a bit out of scope for an elder plugin. I mean, it feels like it can be a separate tool which works with any web project,  with a plugin/wrapper calling this tool through elder hooks.

Anyway, totally in to test and contribute to dev if I have room for it in my schedule.

**Reply by nickreese** [reply]
Agreed. :) Have had it built since launching Elder.js. Just haven’t had s chance to package it.

**Comment by apop880** [comment]
Sounds great, and I'd definitely be up for helping to test.

A good way to set default tags might be a good feature as well. For example, for my Open Graph images, my blog posts right now are set up to use the featured photo, but not all of my other pages have a feature photo. I created a basic logo image that I added for my Home.svelte template, but none of my other routes have anything set up. It would be nice to set up a fallback image (or any other relevant OG tag) that would get written into the head if there isn't one already defined.

**Reply by nickreese** [reply]
Here is a decent library for setting fallbacks... I haven't tested it though: https://github.com/artiebits/svelte-seo

**Comment by nickreese** [comment]
@apop880 @kiuKisas @versysmedia Shipped an alpha version today: https://github.com/Elderjs/plugins/tree/master/packages/seo-check

Let me know your thoughts. Still rough around the edges, but a step in the right direction.

**Comment by nickreese** [comment]
<img width="1143" alt="image" src="https://user-images.githubusercontent.com/3580879/118575127-be6e4b80-b753-11eb-9fff-26498e4ad76f.png">

**Comment by nickreese** [comment]
A more mature version of the plugin is live. It now test 50 common SEO issues: https://github.com/Elderjs/plugins/tree/master/packages/seo-check
